File: scripts/lib/surge_compliance.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Surge external RULE-SET compliance — single source of truth.

Lessons from recent incidents (git history May 2026):
- 053a145d: wrongly stripped PROCESS-NAME / URL-REGEX from .list files
- 427a9228: URL-REGEX truncated when https:// treated as // comment
- b87b2e0f: MetaCubeX DOMAIN-REGEX,$ char-split junk from deco lists
- 0af5f592: quoting DOMAIN-REGEX does NOT fix Surge — type is unsupported
- 0f524ff5: Netflix DOMAIN-REGEX must become DOMAIN-KEYWORD for Surge .list
- d804f1d0: IP-CIDR,...,no-resolve was dropped by naive comma split
"""

from __future__ import annotations

import re
from typing import FrozenSet, Optional, Tuple

# https://manual.nssurge.com/rule/ruleset.html — external .list RULE-SET types
SURGE_RULESET_ALLOWED: FrozenSet[str] = frozenset({
    "DOMAIN", "DOMAIN-SUFFIX", "DOMAIN-KEYWORD", "DOMAIN-SET",
    "IP-CIDR", "IP-CIDR6", "IP-ASN", "GEOIP",
    "USER-AGENT", "URL-REGEX", "PROCESS-NAME",
    "DEST-PORT", "SRC-PORT", "IN-PORT",
})

# Clash-only in external Surge .list (always Invalid line)
SURGE_RULESET_FORBIDDEN: FrozenSet[str] = frozenset({"DOMAIN-REGEX", "DOMAIN-WILDCARD"})

PAYLOAD_SAFE_RULE_TYPES: FrozenSet[str] = frozenset({
    "DOMAIN-REGEX", "URL-REGEX", "USER-AGENT", "PROCESS-NAME",
})

INVALID_DOMAIN_REGEX_VALUES: FrozenSet[str] = frozenset({
    "", "$", ",", "-", ".", "2", "6", "]", "[",
})

INVALID_URL_REGEX_VALUES: FrozenSet[str] = frozenset({
    "", "https:", "http:", "https", "http",
    "^https?:", "^https?://", r"^https?:\/\/",
})

STRIP_POLICY_TOKENS: FrozenSet[str] = frozenset({
    "REJECT", "REJECT-DROP", "REJECT-NO-DROP", "DIRECT", "PROXY",
    "REJECT-TINYGIF", "REJECT-IMG",
    "EXTENDED-MATCHING", "PRE-MATCHING", "NO-RESOLVE", "FORCE-CELLULAR",
    "{{{PROXY}}}",
})


def rule_head(line: str) -> str:
    if "," not in line:
        return line.strip().upper()
    return line.split(",", 1)[0].strip().upper()


def strip_inline_comment(line: str) -> str:
    """Strip # and // comments without breaking https:// inside URL-REGEX."""
    line = line.strip()
    if not line:
        return ""
    if line.startswith("//"):
        return ""
    head = rule_head(line)
    if head in PAYLOAD_SAFE_RULE_TYPES:
        return line
    if "//" in line:
        line = line.split("//", 1)[0].strip()
    if "#" in line:
        line = line.split("#", 1)[0].strip()
    return line


def split_rule(line: str) -> Optional[Tuple[str, str]]:
    """Return (TYPE, payload) after comment strip; None if not a rule line."""
    line = strip_inline_comment(line)
    if not line or "," not in line:
        return None
    rule_type, payload = line.split(",", 1)
    rule_type = rule_type.strip().upper()
    payload = payload.strip()
    if rule_type.startswith("IP-"):
        payload = payload.split(",", 1)[0].strip()
    if payload.startswith('"') and payload.endswith('"'):
        payload = payload[1:-1]
    return rule_type, payload


def strip_trailing_policy(rule: str) -> str:
    """Remove trailing policy/options; preserve commas inside regex payloads."""
    rule = rule.strip()
    if "," not in rule:
        return rule
    head = rule_head(rule)
    if head in PAYLOAD_SAFE_RULE_TYPES:
        rule_type, payload = rule.split(",", 1)
        payload = payload.strip()
        while "," in payload:
            tail = payload.rsplit(",", 1)[-1].strip().upper()
            if tail in STRIP_POLICY_TOKENS or tail.startswith("UPDATE-INTERVAL="):
                payload = payload.rsplit(",", 1)[0].strip()
            else:
                break
        return f"{rule_type.strip()},{payload}"
    parts = [p.strip() for p in rule.split(",")]
    while len(parts) > 2 and (
        parts[-1].upper() in STRIP_POLICY_TOKENS
        or parts[-1].upper().startswith("UPDATE-INTERVAL=")
    ):
        parts = parts[:-1]
    return ",".join(parts)


def is_invalid_domain_regex_payload(payload: str) -> bool:
    return len(payload.strip()) < 2 or payload in INVALID_DOMAIN_REGEX_VALUES


def is_invalid_url_regex_payload(payload: str) -> bool:
    val = payload.strip()
    if len(val) < 3:
        return True
    return val.lower() in INVALID_URL_REGEX_VALUES


def convert_domain_regex_for_surge(rule: str) -> str:
    """Convert Clash DOMAIN-REGEX → Surge-safe DOMAIN-KEYWORD; '' if unmappable."""
    if not rule.upper().startswith("DOMAIN-REGEX,"):
        return ""
    payload = rule.split(",", 1)[1].strip().strip('"')
    if is_invalid_domain_regex_payload(payload):
        return ""

    literal = payload.replace(r"\.", ".")

    m = re.match(r"^\(\^\|\.\)(.+?)-\.\+\.", literal)
    if m:
        return f"DOMAIN-KEYWORD,{m.group(1)}"

    m = re.match(r"^(?:\(\^\|\.\)|\^)([a-zA-Z0-9][-a-zA-Z0-9.]*)$", literal)
    if m:
        return f"DOMAIN-KEYWORD,{m.group(1)}"

    m = re.match(r"^\.\+(.+)$", literal)
    if m and len(m.group(1)) >= 3:
        return f"DOMAIN-KEYWORD,{m.group(1)}"

    return ""


def format_url_regex_for_surge(payload: str) -> str:
    """Quote URL-REGEX when payload contains commas (Surge .list safety)."""
    if "," in payload or " " in payload:
        return f'URL-REGEX,"{payload}"'
    return f"URL-REGEX,{payload}"


def validate_surge_ruleset_line(line: str, *, allow_domain_regex: bool = False) -> Optional[str]:
    """Return error message if line is invalid for Surge external .list; else None."""
    s = line.strip()
    if not s or s.startswith("#"):
        return None

    parsed = split_rule(s)
    if not parsed:
        return "missing rule type or payload"

    rule_type, payload = parsed

    if rule_type in SURGE_RULESET_FORBIDDEN and not allow_domain_regex:
        return f"{rule_type} is not supported in Surge external RULE-SET"

    if rule_type not in SURGE_RULESET_ALLOWED and rule_type not in SURGE_RULESET_FORBIDDEN:
        return f"unknown rule type {rule_type}"

    if rule_type == "URL-REGEX":
        raw = s.split(",", 1)[1].strip()
        quoted = raw.startswith('"') and raw.endswith('"')
        body = raw[1:-1] if quoted else raw
        if "," in body and not quoted:
            return "URL-REGEX payload with comma must be double-quoted"
        if is_invalid_url_regex_payload(body):
            return f"truncated/invalid URL-REGEX {body!r}"
        try:
            re.compile(body)
        except re.error as exc:
            return f"URL-REGEX compile error: {exc}"

    if rule_type == "DOMAIN-REGEX" and allow_domain_regex:
        if is_invalid_domain_regex_payload(payload):
            return f"invalid DOMAIN-REGEX payload {payload!r}"

    return None

File: create/hub/surge_compliance.go
package hub

import (
	"fmt"
	"regexp"
	"strings"
)

var SURGE_RULESET_ALLOWED = map[string]bool{
	"DOMAIN": true, "DOMAIN-SUFFIX": true, "DOMAIN-KEYWORD": true, "DOMAIN-SET": true,
	"IP-CIDR": true, "IP-CIDR6": true, "IP-ASN": true, "GEOIP": true,
	"USER-AGENT": true, "URL-REGEX": true, "PROCESS-NAME": true,
	"DEST-PORT": true, "SRC-PORT": true, "IN-PORT": true,
}

var SURGE_RULESET_FORBIDDEN = map[string]bool{
	"DOMAIN-REGEX": true, "DOMAIN-WILDCARD": true,
}

var PAYLOAD_SAFE_RULE_TYPES = map[string]bool{
	"DOMAIN-REGEX": true, "URL-REGEX": true, "USER-AGENT": true, "PROCESS-NAME": true,
}

var INVALID_DOMAIN_REGEX_VALUES = map[string]bool{
	"": true, "$": true, ",": true, "-": true, ".": true, "2": true, "6": true, "]": true, "[": true,
}

var INVALID_URL_REGEX_VALUES = map[string]bool{
	"": true, "https:": true, "http:": true, "https": true, "http": true,
	"^https?:": true, "^https?://": true, `^https?:\/\/`: true,
}

var STRIP_POLICY_TOKENS = map[string]bool{
	"REJECT": true, "REJECT-DROP": true, "REJECT-NO-DROP": true, "DIRECT": true, "PROXY": true,
	"REJECT-TINYGIF": true, "REJECT-IMG": true,
	"EXTENDED-MATCHING": true, "PRE-MATCHING": true, "NO-RESOLVE": true, "FORCE-CELLULAR": true,
	"{{{PROXY}}}": true,
}

func ruleHead(line string) string {
	if !strings.Contains(line, ",") {
		return strings.ToUpper(strings.TrimSpace(line))
	}
	parts := strings.SplitN(line, ",", 2)
	return strings.ToUpper(strings.TrimSpace(parts[0]))
}

func StripInlineComment(line string) string {
	line = strings.TrimSpace(line)
	if line == "" || strings.HasPrefix(line, "//") {
		return ""
	}
	head := ruleHead(line)
	if PAYLOAD_SAFE_RULE_TYPES[head] {
		return line
	}
	if strings.Contains(line, "//") {
		line = strings.TrimSpace(strings.SplitN(line, "//", 2)[0])
	}
	if strings.Contains(line, "#") {
		line = strings.TrimSpace(strings.SplitN(line, "#", 2)[0])
	}
	return line
}

func splitRule(line string) (string, string, bool) {
	line = StripInlineComment(line)
	if line == "" || !strings.Contains(line, ",") {
		return "", "", false
	}
	parts := strings.SplitN(line, ",", 2)
	ruleType := strings.ToUpper(strings.TrimSpace(parts[0]))
	payload := strings.TrimSpace(parts[1])
	if strings.HasPrefix(ruleType, "IP-") {
		payload = strings.TrimSpace(strings.SplitN(payload, ",", 2)[0])
	}
	if strings.HasPrefix(payload, "\"") && strings.HasSuffix(payload, "\"") && len(payload) >= 2 {
		payload = payload[1 : len(payload)-1]
	}
	return ruleType, payload, true
}

func StripTrailingPolicy(rule string) string {
	rule = strings.TrimSpace(rule)
	if !strings.Contains(rule, ",") {
		return rule
	}
	head := ruleHead(rule)
	if PAYLOAD_SAFE_RULE_TYPES[head] {
		parts := strings.SplitN(rule, ",", 2)
		ruleType := strings.TrimSpace(parts[0])
		payload := strings.TrimSpace(parts[1])
		for strings.Contains(payload, ",") {
			tailParts := strings.Split(payload, ",")
			tail := strings.ToUpper(strings.TrimSpace(tailParts[len(tailParts)-1]))
			if STRIP_POLICY_TOKENS[tail] || strings.HasPrefix(tail, "UPDATE-INTERVAL=") {
				payload = strings.TrimSpace(strings.Join(tailParts[:len(tailParts)-1], ","))
			} else {
				break
			}
		}
		return ruleType + "," + payload
	}

	parts := strings.Split(rule, ",")
	var trimmed []string
	for _, p := range parts {
		trimmed = append(trimmed, strings.TrimSpace(p))
	}
	for len(trimmed) > 2 {
		tail := strings.ToUpper(trimmed[len(trimmed)-1])
		if STRIP_POLICY_TOKENS[tail] || strings.HasPrefix(tail, "UPDATE-INTERVAL=") {
			trimmed = trimmed[:len(trimmed)-1]
		} else {
			break
		}
	}
	return strings.Join(trimmed, ",")
}

func IsInvalidDomainRegexPayload(payload string) bool {
	if len(strings.TrimSpace(payload)) < 2 {
		return true
	}
	return INVALID_DOMAIN_REGEX_VALUES[payload]
}

func IsInvalidUrlRegexPayload(payload string) bool {
	val := strings.TrimSpace(payload)
	if len(val) < 3 {
		return true
	}
	return INVALID_URL_REGEX_VALUES[strings.ToLower(val)]
}

func ConvertDomainRegexForSurge(rule string) string {
	if !strings.HasPrefix(strings.ToUpper(rule), "DOMAIN-REGEX,") {
		return ""
	}
	parts := strings.SplitN(rule, ",", 2)
	payload := strings.Trim(strings.TrimSpace(parts[1]), "\"")
	if IsInvalidDomainRegexPayload(payload) {
		return ""
	}

	literal := strings.ReplaceAll(payload, "\\.", ".")

	m1 := regexp.MustCompile(`^\(\^\|\.\)(.+?)-\.\+\.`)
	match1 := m1.FindStringSubmatch(literal)
	if match1 != nil {
		return "DOMAIN-KEYWORD," + match1[1]
	}

	m2 := regexp.MustCompile(`^(?:\(\^\|\.\)|\^)([a-zA-Z0-9][-a-zA-Z0-9.]*)$`)
	match2 := m2.FindStringSubmatch(literal)
	if match2 != nil {
		return "DOMAIN-KEYWORD," + match2[1]
	}

	m3 := regexp.MustCompile(`^\.\+(.+)$`)
	match3 := m3.FindStringSubmatch(literal)
	if match3 != nil && len(match3[1]) >= 3 {
		return "DOMAIN-KEYWORD," + match3[1]
	}

	return ""
}

func FormatUrlRegexForSurge(payload string) string {
	if strings.Contains(payload, ",") || strings.Contains(payload, " ") {
		return fmt.Sprintf("URL-REGEX,\"%s\"", payload)
	}
	return "URL-REGEX," + payload
}

func ValidateSurgeRulesetLine(line string, allowDomainRegex bool) *string {
	s := strings.TrimSpace(line)
	if s == "" || strings.HasPrefix(s, "#") {
		return nil
	}

	ruleType, payload, ok := splitRule(s)
	if !ok {
		errStr := "missing rule type or payload"
		return &errStr
	}

	if SURGE_RULESET_FORBIDDEN[ruleType] && !allowDomainRegex {
		errStr := fmt.Sprintf("%s is not supported in Surge external RULE-SET", ruleType)
		return &errStr
	}

	if !SURGE_RULESET_ALLOWED[ruleType] && !SURGE_RULESET_FORBIDDEN[ruleType] {
		errStr := fmt.Sprintf("unknown rule type %s", ruleType)
		return &errStr
	}

	if ruleType == "URL-REGEX" {
		raw := strings.TrimSpace(strings.SplitN(s, ",", 2)[1])
		quoted := strings.HasPrefix(raw, "\"") && strings.HasSuffix(raw, "\"")
		body := raw
		if quoted && len(raw) >= 2 {
			body = raw[1 : len(raw)-1]
		}
		if strings.Contains(body, ",") && !quoted {
			errStr := "URL-REGEX payload with comma must be double-quoted"
			return &errStr
		}
		if IsInvalidUrlRegexPayload(body) {
			errStr := fmt.Sprintf("truncated/invalid URL-REGEX %q", body)
			return &errStr
		}
		_, err := regexp.Compile(body)
		if err != nil {
			errStr := fmt.Sprintf("URL-REGEX compile error: %v", err)
			return &errStr
		}
	}

	if ruleType == "DOMAIN-REGEX" && allowDomainRegex {
		if IsInvalidDomainRegexPayload(payload) {
			errStr := fmt.Sprintf("invalid DOMAIN-REGEX payload %q", payload)
			return &errStr
		}
	}

	return nil
}
